File: src/main/java/com/kingmang/lazurite/exceptions/parser/ParseErrors.java
package com.kingmang.lazurite.exceptions.parser;

import com.kingmang.lazurite.console.Console;
import com.kingmang.lazurite.exceptions.parser.ParseError;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public final class ParseErrors implements Iterable<ParseError> {

    private final List<ParseError> errors;

    public ParseErrors() {
        errors = new ArrayList<>();
    }
    
    public void clear() {
        errors.clear();
    }
    
    public void add(Exception ex, int line) {
        errors.add(new ParseError(line, ex));
    }
    
    public boolean hasErrors() {
        return !errors.isEmpty();
    }

    @Override
    public Iterator<ParseError> iterator() {
        return errors.iterator();
    }

    @Override
    public String toString() {
        final StringBuilder result = new StringBuilder();
        for (ParseError error : errors) {
            result.append(error).append(Console.newline());
        }
        return result.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/exceptions/parser/ParseErrors.kt
package com.kingmang.lazurite.exceptions.parser

import com.kingmang.lazurite.console.Console.newline

class ParseErrors : Iterable<ParseError?> {
    private val errors: MutableList<ParseError> = ArrayList()

    fun clear() =
        errors.clear()

    fun add(ex: Exception?, line: Int) {
        errors.add(ParseError(line, ex!!))
    }

    fun hasErrors(): Boolean =
        errors.isNotEmpty()

    override fun iterator(): MutableIterator<ParseError> =
        errors.iterator()

    override fun toString(): String {
        val result = StringBuilder()
        errors.forEach { result.append(it).append(newline()) }
        return result.toString()
    }
}
